porch/apiserver/pkg/apiserver/apiserver.go

```diff
@@ -168,13 +168,21 @@ func (c completedConfig) New() (*PorchServer, error) {
 	renderer := kpt.NewRenderer()
 
 	cache := cache.NewCache(c.ExtraConfig.CacheDirectory, credentialResolver)
-	cad, err := engine.NewCaDEngine(
+	options := []engine.EngineOption{
 		engine.WithCache(cache),
-		engine.WithGRPCFunctionRuntime(c.ExtraConfig.FunctionRunnerAddress),
 		engine.WithCredentialResolver(credentialResolver),
 		engine.WithRenderer(renderer),
 		engine.WithReferenceResolver(referenceResolver),
-	)
+	}
+
+	if c.ExtraConfig.FunctionRunnerAddress != "" {
+		options = append(options, engine.WithGRPCFunctionRuntime(c.ExtraConfig.FunctionRunnerAddress))
+	} else {
+		ns := "porch-functions-system"
+		options = append(options, engine.WithKubeFunctionRuntime(coreClient, ns))
```

Comment:
this seems to read as though the runtimes are mutually exclusive. We still need high performance runtime for cached functions.

(since there appears to be no example of what the Kube function pod looks like, it is hard to tell what expected performance they will have)

Reply:
They are currently mutually exclusive, and I changed this PR so that they aren't enabled by default (I think of it as a first step to start investigating this path).

I can add a "try multiple engines" runtime, but we should try to figure out what we need.

There are some example servers on my experimental branch, and they use [something similar to your code](https://github.com/justinsb/kpt/blob/main_integration/porch/functions/pkg/yo/krmfunction/server.go), except the functions [implement the server directly](https://github.com/justinsb/kpt/blob/main_integration/porch/config/samples/functions/main.go) rather than being invoked as binaries.

Still lots of experiments to be done here, e.g. should we multiplex functions into an image?  How long should we run a pod for?  Is the performance good enough?  My impression is that the performance is good enough, but I'm also (manually) pre-launching all the pods!

Reply:
Experimental "multiple-engines" runtime, if we need it: https://github.com/GoogleContainerTools/kpt/pull/2937